File: android/app/src/main/java/com/soulo/app/models/CognitiveDrift.kt
package com.soulo.app.models

import kotlinx.serialization.Serializable

@Serializable
enum class DriftSignificance {
    none, mild, moderate, significant, critical;

    val score: Int get() = when (this) {
        none -> 0; mild -> 25; moderate -> 50; significant -> 75; critical -> 95
    }
}

@Serializable
data class CognitiveDrift(
    val id: String,
    val windowStart: Long,
    val windowEnd: Long,
    val metric: BiomarkerMetric,
    val delta: Double,
    val meanEarly: Double,
    val meanLate: Double,
    val direction: TrendDirection,
    val significance: DriftSignificance,
    val correlation: Map<EmotionType, Double>,
    val detectedAt: Long
) {
    val description: String
        get() {
            val dir = if (direction == TrendDirection.increasing) "increased" else "decreased"
            return "${metric.name} $dir by ${String.format("%.2f", kotlin.math.abs(delta))}"
        }
}

@Serializable
data class DriftReport(
    val id: String,
    val overallSignificance: DriftSignificance,
    val drifts: List<CognitiveDrift>,
    val trendShiftCount: Int,
    val emotionalCorrelation: Map<EmotionType, Double>,
    val gradualVsSudden: DriftType,
    val summary: String,
    val detectedAt: Long
) {
    @Serializable
    enum class DriftType { gradual, sudden, mixed }
}

@Serializable
data class DailyPattern(
    val metric: String,
    val morningAvg: Double, val afternoonAvg: Double,
    val eveningAvg: Double, val nightAvg: Double,
    val morningSampleCount: Int, val afternoonSampleCount: Int,
    val eveningSampleCount: Int, val nightSampleCount: Int,
    val hasSignificantPattern: Boolean
) {
    val peakPeriod: String
        get() = listOf(
            morningAvg to "morning", afternoonAvg to "afternoon",
            eveningAvg to "evening", nightAvg to "night"
        ).maxByOrNull { it.first }?.second ?: "unknown"

    val troughPeriod: String
        get() = listOf(
            morningAvg to "morning", afternoonAvg to "afternoon",
            eveningAvg to "evening", nightAvg to "night"
        ).minByOrNull { it.first }?.second ?: "unknown"
}

File: Soulo/Models/CognitiveDrift.swift
import Foundation

struct CognitiveDrift: Identifiable, Codable, Sendable {
    let id: String
    let windowStart: TimeInterval
    let windowEnd: TimeInterval
    let metric: BiomarkerMetric
    let delta: Double
    let meanEarly: Double
    let meanLate: Double
    let direction: TrendDirection
    let significance: DriftSignificance
    let correlation: [EmotionType: Double]
    let detectedAt: TimeInterval

    var description: String {
        let dir = direction == .increasing ? "increased" : "decreased"
        return "\(metric.rawValue) \(dir) by \(String(format: "%.2f", abs(delta)))"
    }
}

enum DriftSignificance: String, Codable, Comparable {
    case none
    case mild
    case moderate
    case significant
    case critical

    var score: Int {
        switch self {
        case .none: return 0
        case .mild: return 25
        case .moderate: return 50
        case .significant: return 75
        case .critical: return 95
        }
    }

    static func < (lhs: DriftSignificance, rhs: DriftSignificance) -> Bool {
        lhs.score < rhs.score
    }
}

struct DriftReport: Codable, Sendable {
    let id: String
    let overallSignificance: DriftSignificance
    let drifts: [CognitiveDrift]
    let trendShiftCount: Int
    let emotionalCorrelation: [EmotionType: Double]
    let gradualVsSudden: DriftType
    let summary: String
    let detectedAt: TimeInterval

    enum DriftType: String, Codable {
        case gradual
        case sudden
        case mixed
    }
}

struct DailyPattern: Codable, Sendable {
    let metric: String
    let morningAvg: Double
    let afternoonAvg: Double
    let eveningAvg: Double
    let nightAvg: Double
    let morningSampleCount: Int
    let afternoonSampleCount: Int
    let eveningSampleCount: Int
    let nightSampleCount: Int
    let hasSignificantPattern: Bool

    var peakPeriod: String {
        let vals = [(morningAvg, "morning"), (afternoonAvg, "afternoon"),
                    (eveningAvg, "evening"), (nightAvg, "night")]
        return vals.max { $0.0 < $1.0 }?.1 ?? "unknown"
    }

    var troughPeriod: String {
        let vals = [(morningAvg, "morning"), (afternoonAvg, "afternoon"),
                    (eveningAvg, "evening"), (nightAvg, "night")]
        return vals.min { $0.0 < $1.0 }?.1 ?? "unknown"
    }
}
